streams/jvm/src/main/scala/zio/stream/platform.scala

```diff
@@ -236,4 +239,81 @@ trait ZStreamPlatformSpecificConstructors { self: ZStream.type =>
    */
   final def fromJavaStreamTotal[A](stream: => ju.stream.Stream[A]): ZStream[Any, Nothing, A] =
     ZStream.fromJavaIteratorTotal(stream.iterator())
+
+  /**
+   * Create a stream of accepted connection from server socket
+   * Emit socket `Connection` from which you can read / write and ensure it is closed after it is used
+   */
+  def fromSocketServer(
+    port: Int,
+    host: Option[String] = None
+  ): ZStream[Blocking, Throwable, Managed[Throwable, Connection]] =
+    for {
+      server <- ZStream.managed(ZManaged.fromAutoCloseable(Task {
+                 AsynchronousServerSocketChannel
+                   .open()
+                   .bind(
+                     host.fold(new InetSocketAddress(port))(new InetSocketAddress(_, port))
+                   )
+               }))
+
+      conn <- ZStream.unfoldM(server) { s =>
+               IO.effectAsync[Throwable, Connection] { callback =>
+                   s.accept(
+                     null,
+                     new CompletionHandler[AsynchronousSocketChannel, Void]() {
+                       self =>
+                       override def completed(result: AsynchronousSocketChannel, attachment: Void): Unit =
+                         callback(ZIO.succeed(new Connection(result)))
+
+                       override def failed(exc: Throwable, attachment: Void): Unit = callback(ZIO.fail(exc))
+                     }
+                   )
+                 }
+                 .map(c => Option(c -> s))
+             }
+
+    } yield Managed.make(ZIO.succeed(conn))(c => ZIO.effectTotal(c.close()))
+
+  private[zio] class Connection(socket: AsynchronousSocketChannel) {
```

Comment:
This should probably be public.

Reply:
Let's give this a companion object with a make function that returns a `ZManaged`.